docs/proposals/resource-catalog.md

```diff
@@ -0,0 +1,222 @@
+
+
+We can reshape the catalog resource to more easily express them, secure them, and claim them.
+
+# Goals
+ 1.  cluster-admin should control which catalog exist.
+ 2.  cluster-admin should be able/disable per-catalog exposure on a per-project granularity.
+ 3.  project-admin should be able to delegate per-catalog exposure.
+ 4.  project-editor should be able to claim an entry from the catalog.
+ 5.  ResourceCatalogClaims should be sync-able, but not require syncing.
+ 6.  project-admin should be able to control the shape of claim realization and prevent unwanted injection.
+ 7.  ResourceCatalogEntries should not be restricted to specific claim realization mechanism.
+
+To achieve the ACL related goals, we can create namespaced resources that are selected by label at a cluster level.
+
+
+# API Objects
+
+## ResourceCatalog
+ResourceCatalog is a cluster-scoped resource.  It has a cluster-admin controlled spec that is used to drive a controller.
+The spec includes a label selector that is used in a list/watch to filter all ResourceCatalogEntries in the cluster.
+The controller will find matching entries and copy a reference into the ResourceCatalog.Status field.  The controller
+should have a plug-point for deciding if a given entry is accepted into a given catalog: `Accept(catalogName, entryNamespace, entryName)`
+is probably sufficient.
+
+```go
+// ResourceCatalog is a cluster-scoped resource with a specified spec and a controller maintained status
+type ResourceCatalog struct {
+	unversioned.TypeMeta
+	api.ObjectMeta
+
+	Spec   ResourceCatalogSpec
+	Status ResourceCatalogStatus
+}
+
+type ResourceCatalogSpec struct {
+	// Selector is the selector to use when gathering the namespace scope `ResourceCatalogEntries` for
+	// the `Status`
+	Selector map[string]string
+}
+
+type ResourceCatalogStatus struct {
+	// Entries is dynamically retrieved by selecting all matching ResourceCatalogEntries from all namespaces
+	// This can be done by a controller using a list/watch on all ResourceCatalogEntries based on the spec.selector.
+	Entries []ResourceCatalogInfo
+}
+
+// ResourceCatalogInfo exists so that the details of how the ResourceCatalogEntry statisfies the request
+// can be properly abstracted from whoever claims it.  It also adds a degree of privacy for the
+// ResourceCatalogEntry
+type ResourceCatalogInfo struct {
+	// Namespace is system controlled.  It is the namespace of the originating ResourceCatalogEntry
+	Namespace string
+	// Name  is system controlled.  It is the name of the ResourceCatalogEntry
+	Name string
+	// UID  is system controlled.  It is the UID of the ResourceCatalogEntry
+	UID string
+	// Generation indicates which level of the Entry was accepted
+	Generation int
+
+	// Description is client controlled.  It describes the resources being provided.
+	Description string
```

Comment:
Do you think it makes sense to include the types & names of resources that will end up being provisioned? This would be so that when you look at a catalog entry, you can see up front that you'll get this secret or that configmap. And if we go that far, maybe we also enumerate the names of the keys too?

What I'm looking for is an easy way for a user to know that when they find an entry in the catalog, they will refer to it using the equivalent of v2's OPENSHIFT_MYSQL_PASSWORD etc.